File: frontend/src/types/centralFinanceira.ts
/**
 * Espelha 1:1 os schemas de saída de `app/schemas/central_financeira.py` e
 * das entidades que ela reaproveita (`ContaRead`, `CartaoRead`, `FaturaRead`,
 * `FinanciamentoRead`, `EmprestimoRead`, `MetaRead`) — conferido por leitura
 * direta do backend, não da especificação de produto original. Ver
 * docs/analise-arquitetural-dashboard.md, seção 3 e 6.1.
 *
 * Só os campos de leitura (`Read`) — sem `Create`/`Update` aqui, esta etapa
 * é 100% somente-leitura. Quando o CRUD de cada entidade chegar (F6+), os
 * tipos completos nascem em `types/<entidade>.ts` e estes podem ser
 * re-exportados de lá (sinalizado em analise-arquitetural-dashboard.md,
 * seção 6.1) — `ContaRead` já foi migrado nesse sentido na Etapa F6 (CRUD
 * de Conta): a definição completa mora em `types/conta.ts`, aqui é só
 * reexportada para não quebrar nenhum import existente.
 *
 * Todo valor monetário/decimal é `string` (nunca `number`) — mesmo princípio
 * de `analise-arquitetural-frontend.md`, seção 0.
 */
import type {
  Bandeira,
  CategoriaEventoCalendario,
  SistemaAmortizacao,
  StatusContratoCredito,
  StatusFatura,
  TipoEntidadeReferenciavel,
} from "./enums";
import type { ContaRead } from "./conta";
import type { MetaRead } from "./meta";

export type { ContaRead, MetaRead };

// ---- Entidades reaproveitadas (recorte "Read" mínimo usado pela Central) ----

export interface CartaoRead {
  id: number;
  nome: string;
  conta_pagamento_id: number;
  instituicao: string;
  bandeira: Bandeira;
  ultimos_quatro_digitos: string;
  limite: string;
  limite_disponivel: string;
  dia_fechamento: number;
  dia_vencimento: number;
  ativo: boolean;
}

export interface FaturaRead {
  id: number;
  cartao_id: number;
  mes_referencia: string;
  data_fechamento: string;
  data_vencimento: string;
  valor_pago: string;
  valor_total: string;
  status: StatusFatura;
}

interface ContratoCreditoBase {
  id: number;
  descricao: string;
  instituicao_financeira: string;
  numero_contrato: string | null;
  taxa_juros: string;
  sistema_amortizacao: SistemaAmortizacao;
  num_parcelas: number;
  cet: string | null;
  data_inicio: string;
  saldo_devedor: string;
  permite_quitacao_antecipada: boolean;
  status: StatusContratoCredito;
  conta_id: number | null;
  categoria_id: number | null;
  parcelas_pagas: number;
  parcelas_restantes: number;
  valor_total_pago: string;
  proxima_parcela_data: string | null;
  proxima_parcela_valor: string | null;
}

export interface FinanciamentoResumo extends ContratoCreditoBase {
  valor_financiado: string;
  valor_entrada: string | null;
  bem_financiado: string | null;
}

export interface EmprestimoResumo extends ContratoCreditoBase {
  valor_liberado: string;
  finalidade: string | null;
}

// ---- Schemas de saída da Central Financeira (11 endpoints) ----
// `MetaRead` é reexportado de `./meta` (mesma migração já feita para
// `ContaRead` na Etapa F6): o backend de `/central-financeira/metas`
// (`ProgressoMetasRead`) reusa literalmente `app.schemas.meta.MetaRead`,
// então o tipo completo (incluindo os campos do Refinamento de Metas —
// `frequencia_contribuicao`, `situacao_planejamento`, etc.) já chega pronto
// aqui sem precisar de nenhuma alteração no backend.

export interface ContaSaldoResumo {
  id: number;
  nome: string;
  saldo_atual: string;
}

export interface SaldoConsolidadoRead {
  saldo_total: string;
  contas: ContaSaldoResumo[];
}

export interface ResumoContasRead {
  contas: ContaRead[];
}

export interface ResumoCartoesRead {
  cartoes: CartaoRead[];
  total_utilizado: string;
}

/** `/central-financeira/cartoes/agregado` — panorama do "Dashboard de
 * Cartões" (Sprint de Refinamento Premium, item 3): deliberadamente sem
 * nenhum gráfico por cartão individual (isso é `/cartoes/:id`). Todo campo
 * já vem somado/calculado pelo backend a partir de `CartaoRead`/`FaturaRead`
 * — este tipo só espelha a resposta, nenhum cálculo aqui. */
export interface UsoCartao {
  cartao_id: number;
  nome: string;
  percentual_usado: string;
}

export interface ProximoVencimentoFatura {
  cartao_id: number;
  cartao_nome: string;
  fatura_id: number;
  data_vencimento: string;
  valor_total: string;
}

export interface ResumoCartoesAgregadoRead {
  limite_total: string;
  limite_disponivel_total: string;
  limite_usado_total: string;
  percentual_usado_geral: string;
  quantidade_cartoes: number;
  faturas_em_aberto: number;
  proximos_vencimentos: ProximoVencimentoFatura[];
  distribuicao_uso: UsoCartao[];
}

export interface ResumoFaturasRead {
  faturas: FaturaRead[];
}

export interface ResumoFinanciamentosRead {
  financiamentos: FinanciamentoResumo[];
}

export interface ResumoEmprestimosRead {
  emprestimos: EmprestimoResumo[];
}

export interface ProgressoMetasRead {
  metas: MetaRead[];
}

export interface EventoAgenda {
  data: string;
  descricao: string;
  valor: string;
  origem_tipo: TipoEntidadeReferenciavel;
  origem_id: number;
}

export interface AgendaFinanceiraRead {
  eventos: EventoAgenda[];
}

export interface ResumoFinanceiroRead {
  ano: number;
  mes: number;
  saldo_total: string;
  entradas_mes: string;
  saidas_mes: string;
  fluxo_caixa_mes: string;
  patrimonio_liquido: string;
}

export interface VisaoMensalRead {
  ano: number;
  mes: number;
  entradas: string;
  saidas: string;
  fluxo_caixa: string;
}

export interface IndicadoresGeraisRead {
  contas_ativas: number;
  cartoes_ativos: number;
  faturas_em_aberto: number;
  financiamentos_ativos: number;
  emprestimos_ativos: number;
  metas_ativas: number;
  percentual_medio_metas: string;
  parcelas_atrasadas: number;
}

/** `GET /central-financeira/calendario` — endpoint IRMÃO de `agenda` acima
 * (não substitui): escopo é o MÊS inteiro (`ano`/`mes`), qualquer status, e
 * inclui Transferência/Meta/fechamento de fatura, que `agenda` nunca
 * incluiu. Ver docs/analise-arquitetural-transferencias-frontend.md. */
export interface EventoCalendario {
  data: string;
  descricao: string;
  valor: string;
  categoria: CategoriaEventoCalendario;
  origem_tipo: TipoEntidadeReferenciavel;
  origem_id: number;
  status: string | null;
  /** Expansão de Contas Recorrentes (2026-07-20): `true` só para
   * ocorrências FUTURAS projetadas de um template ativo (horizonte de 90
   * dias, nunca persistidas) — a UI as renderiza como previsão (estilo
   * tracejado/atenuado), nunca como história. */
  previsto: boolean;
}

export interface CalendarioFinanceiroRead {
  ano: number;
  mes: number;
  eventos: EventoCalendario[];
}

/** `GET /central-financeira/atividades` — Central de Atividades (Sprint de
 * Refinamento Premium, item 17): feed cronológico de "o que aconteceu
 * recentemente", combinando Transação/Transferência/Meta concluída.
 * `data_hora` é sempre um datetime ISO (mesmo para Meta, cujo backend
 * normaliza `concluida_em` para meia-noite) — único jeito de ordenar as 3
 * fontes de forma consistente. `origem_tipo`/`origem_id` reaproveitam o
 * mesmo discriminador de `EventoCalendario`/`EventoAgenda`
 * (`TipoEntidadeReferenciavel`), então o mesmo mapa `ICONE_POR_ORIGEM`/
 * `ROTA_POR_ORIGEM` de `lib/origemNavegacao.ts` serve aqui também. */
export interface AtividadeRecente {
  data_hora: string;
  descricao: string;
  valor: string | null;
  origem_tipo: TipoEntidadeReferenciavel;
  origem_id: number;
}

export interface CentralAtividadesRead {
  atividades: AtividadeRecente[];
}

/** Etapa de Gráficos (docs/analise-arquitetural-graficos.md) — os 2 novos
 * endpoints agregam exatamente os 4 gráficos da primeira etapa + o 5º
 * (distribuição do saldo atual, que reaproveita `SaldoConsolidadoRead`
 * acima, sem tipo próprio). */
export interface PontoTendenciaMensal {
  ano: number;
  mes: number;
  saldo_total: string;
  entradas: string;
  saidas: string;
}

export interface GraficosTendenciasRead {
  meses: PontoTendenciaMensal[];
}

export interface GastoPorCategoria {
  categoria_id: number | null;
  categoria_nome: string;
  categoria_cor: string | null;
  categoria_icone: string | null;
  total: string;
}

export interface GastoPorCartao {
  cartao_id: number;
  cartao_nome: string;
  total: string;
}

export interface GraficosPeriodoRead {
  ano: number;
  mes: number;
  gastos_por_categoria: GastoPorCategoria[];
  gastos_por_cartao: GastoPorCartao[];
}

File: backend/app/schemas/central_financeira.py
"""Schemas de saída da Central Financeira: `app/api/routes/central_financeira.py`.

Camada 100% somente-leitura - não existe `Create`/`Update` aqui, e nenhum
schema tem `usuario_id` (a Central nunca aceita esse campo do cliente,
mesmo padrão de toda rota autenticada do projeto). Todo valor exposto aqui
já foi calculado por um Service de domínio existente (`ContaRead.saldo_atual`,
`CartaoRead.limite_disponivel`, `FaturaRead.valor_total`/`status`,
`MetaRead.valor_acumulado`/`percentual`, `FinanciamentoRead`/
`EmprestimoRead.saldo_devedor`) - os schemas abaixo só agrupam/formatam,
nunca recalculam. Ver docs/analise-arquitetural-central-financeira.md.
"""
from datetime import date, datetime
from decimal import Decimal

from app.models.enums import CategoriaEventoCalendario, TipoEntidadeReferenciavel
from app.schemas.base import OrmBaseModel
from app.schemas.cartao import CartaoRead
from app.schemas.conta import ContaRead
from app.schemas.emprestimo import EmprestimoRead
from app.schemas.fatura import FaturaRead
from app.schemas.financiamento import FinanciamentoRead
from app.schemas.meta import MetaRead


class ContaSaldoResumo(OrmBaseModel):
    """Recorte mínimo de Conta usado dentro de `SaldoConsolidadoRead` -
    `ContaRead` completo já existe e é o que `GET /central-financeira/contas`
    devolve; aqui só o necessário para a foto de saldo consolidado."""

    id: int
    nome: str
    saldo_atual: Decimal


class SaldoConsolidadoRead(OrmBaseModel):
    saldo_total: Decimal
    contas: list[ContaSaldoResumo]


class ResumoContasRead(OrmBaseModel):
    contas: list[ContaRead]


class ResumoCartoesRead(OrmBaseModel):
    cartoes: list[CartaoRead]
    total_utilizado: Decimal


class UsoCartao(OrmBaseModel):
    """Um item da distribuição de uso por cartão, dentro de
    `ResumoCartoesAgregadoRead` - `percentual_usado` é lido de
    `CartaoRead.limite`/`limite_disponivel` (já calculados por
    `CartaoService`), nunca recalculado aqui."""

    cartao_id: int
    nome: str
    percentual_usado: Decimal


class ProximoVencimentoFatura(OrmBaseModel):
    cartao_id: int
    cartao_nome: str
    fatura_id: int
    data_vencimento: date
    valor_total: Decimal


class ResumoCartoesAgregadoRead(OrmBaseModel):
    """Panorama agregado para o "Dashboard de Cartões" (Sprint de
    Refinamento Premium, item 3) - deliberadamente SEM nenhum gráfico por
    cartão individual (isso é `/cartoes/:id`). Todo valor é soma/contagem
    sobre a lista de `CartaoRead`/`FaturaRead` que os Services de domínio
    já devolvem calculados - mesmas 3 regras estruturais do topo deste
    arquivo."""

    limite_total: Decimal
    limite_disponivel_total: Decimal
    limite_usado_total: Decimal
    percentual_usado_geral: Decimal
    quantidade_cartoes: int
    faturas_em_aberto: int
    proximos_vencimentos: list[ProximoVencimentoFatura]
    distribuicao_uso: list[UsoCartao]


class ResumoFaturasRead(OrmBaseModel):
    faturas: list[FaturaRead]


class FinanciamentoResumo(FinanciamentoRead):
    """`FinanciamentoRead` + métricas de acompanhamento que a Central
    orquestra a partir de `TransacaoService` (parcelas do contrato) - nenhuma
    delas é recalculada aqui, só lida via `TransacaoService.listar`."""

    parcelas_pagas: int
    parcelas_restantes: int
    valor_total_pago: Decimal
    proxima_parcela_data: date | None
    proxima_parcela_valor: Decimal | None


class ResumoFinanciamentosRead(OrmBaseModel):
    financiamentos: list[FinanciamentoResumo]


class EmprestimoResumo(EmprestimoRead):
    parcelas_pagas: int
    parcelas_restantes: int
    valor_total_pago: Decimal
    proxima_parcela_data: date | None
    proxima_parcela_valor: Decimal | None


class ResumoEmprestimosRead(OrmBaseModel):
    emprestimos: list[EmprestimoResumo]


class ProgressoMetasRead(OrmBaseModel):
    metas: list[MetaRead]


class EventoAgenda(OrmBaseModel):
    """Um evento futuro na Agenda Financeira - projeção somente-leitura, sem
    tabela própria (ver docs/analise-arquitetural-central-financeira.md,
    seção 4). `origem_tipo` reaproveita `TipoEntidadeReferenciavel` (já
    existente para `Alerta`/`Anexo`) em vez de um enum novo."""

    data: date
    descricao: str
    valor: Decimal
    origem_tipo: TipoEntidadeReferenciavel
    origem_id: int


class AgendaFinanceiraRead(OrmBaseModel):
    eventos: list[EventoAgenda]


class ResumoFinanceiroRead(OrmBaseModel):
    ano: int
    mes: int
    saldo_total: Decimal
    entradas_mes: Decimal
    saidas_mes: Decimal
    fluxo_caixa_mes: Decimal
    patrimonio_liquido: Decimal


class VisaoMensalRead(OrmBaseModel):
    ano: int
    mes: int
    entradas: Decimal
    saidas: Decimal
    fluxo_caixa: Decimal


class IndicadoresGeraisRead(OrmBaseModel):
    contas_ativas: int
    cartoes_ativos: int
    faturas_em_aberto: int
    financiamentos_ativos: int
    emprestimos_ativos: int
    metas_ativas: int
    percentual_medio_metas: Decimal
    parcelas_atrasadas: int


class EventoCalendario(OrmBaseModel):
    """Um evento do Calendário Financeiro mensal
    (`GET /central-financeira/calendario`) - endpoint NOVO, irmão de
    `EventoAgenda`/`AgendaFinanceiraRead` acima, não uma substituição: a
    Agenda (widget do Dashboard) continua servindo "próximos N dias,
    só pendente"; o Calendário serve "o mês inteiro, todo status" (ver
    `CentralFinanceiraService.calendario_financeiro`).

    `categoria` (`CategoriaEventoCalendario`) decide a COR do indicador
    (dot) no calendário. `origem_tipo`/`origem_id` (`TipoEntidadeReferenciavel`,
    o mesmo enum de `EventoAgenda`) decidem para ONDE navegar ao clicar -
    são dois discriminadores independentes de propósito (ver docstring de
    `CategoriaEventoCalendario` em `models/enums.py`).

    `status`/`horario` são só apresentação no Drawer do dia (ex.: "PAGO",
    "PENDENTE", "FECHADA", "ATIVA", "CANCELADA") - texto livre porque cada
    entidade de origem tem seu próprio enum de status (`StatusTransacao`,
    `StatusFatura`, etc.); o Service já lê `.value` antes de devolver, então
    aqui é sempre `str`, nunca um enum específico de uma entidade só.
    """

    data: date
    descricao: str
    valor: Decimal
    categoria: CategoriaEventoCalendario
    origem_tipo: TipoEntidadeReferenciavel
    origem_id: int
    status: str | None = None
    # Expansão de Contas Recorrentes (2026-07-20): True só para
    # ocorrências FUTURAS projetadas de um template ATIVO (horizonte de 90
    # dias, nunca persistidas - ver
    # `ContaRecorrenteService.projetar_ocorrencias`). Default False =
    # mudança aditiva, nenhum evento/consumidor existente muda.
    previsto: bool = False


class CalendarioFinanceiroRead(OrmBaseModel):
    ano: int
    mes: int
    eventos: list[EventoCalendario]


class AtividadeRecente(OrmBaseModel):
    """Um item da Central de Atividades (`GET /central-financeira/atividades`,
    Sprint de Refinamento Premium, item 17) - feed cronológico de "o que
    aconteceu recentemente", combinando Transação/Transferência/Meta
    concluída (ver `CentralFinanceiraService.atividades_recentes`).
    `data_hora` é sempre um `datetime` (mesmo para Meta, cujo
    `concluida_em` é só `date` - o Service normaliza para meia-noite ao
    montar isto), único jeito de ordenar as 3 fontes de forma consistente.
    `origem_tipo`/`origem_id` reaproveitam o mesmo discriminador de
    `EventoCalendario`/`EventoAgenda` (`TipoEntidadeReferenciavel`) - zero
    tipo novo, mesmo mapa de ícone/rota do frontend serve aqui também."""

    data_hora: datetime
    descricao: str
    valor: Decimal | None = None
    origem_tipo: TipoEntidadeReferenciavel
    origem_id: int


class CentralAtividadesRead(OrmBaseModel):
    atividades: list[AtividadeRecente]


class PontoTendenciaMensal(OrmBaseModel):
    """Um mês da série de `GET /central-financeira/graficos/tendencias`
    (docs/analise-arquitetural-graficos.md). `saldo_total` é ACUMULADO
    (saldo de todas as contas ao FINAL deste mês); `entradas`/`saidas` são
    só deste mês (não acumuladas) - ver
    `CentralFinanceiraService.graficos_tendencias`."""

    ano: int
    mes: int
    saldo_total: Decimal
    entradas: Decimal
    saidas: Decimal


class GraficosTendenciasRead(OrmBaseModel):
    meses: list[PontoTendenciaMensal]


class GastoPorCategoria(OrmBaseModel):
    """`categoria_id` pode ser `None` (lançamento sem categoria) -
    `categoria_nome` já vem resolvido como "Sem categoria" nesse caso,
    nunca omitido do gráfico."""

    categoria_id: int | None
    categoria_nome: str
    categoria_cor: str | None
    categoria_icone: str | None
    total: Decimal


class GastoPorCartao(OrmBaseModel):
    cartao_id: int
    cartao_nome: str
    total: Decimal


class GraficosPeriodoRead(OrmBaseModel):
    ano: int
    mes: int
    gastos_por_categoria: list[GastoPorCategoria]
    gastos_por_cartao: list[GastoPorCartao]
